### DISCUSSION TITLE
Unable to setup authentication with own Identity Server 4 provider

### DISCUSSION THREAD

**Discussion by StuFrankish** [requarks/wiki#3450] [Help / Questions]
**Question**
I run a custom OpenID Identity provider running Identity Server 4 - I have created a client entity within it that with what I believe is all the correct information but when I select the option from the login screen to use it, Identity Server returns an `unauthorised_client` error.

The setup within WikiJS for the provider is set to;
Generic OpenID Connect / OAuth2
ClientID: WIKI (matches the client ID in the ClientEntity table and within the client data
ClientSecret: *matched the value in ClientSecrets[0].Value*
Remaining endpoint URLs all match the endpoint URLs in the discovery document of the provider.
The Redirect URI's are all also set correctly in the Client Data within my Identity provider and are coming through in the URL during the login.

**Host Info (please complete the following information):**
OS: [Docker]
Wiki.js version: 2.5.144
Database engine: SQLite

**Comment by StuFrankish** [comment]
I've managed to access the error message returned by the provider.
I'm getting the message "Invalid grant type for client".

My configuration data for the client has allowed grant types containing only "hybrid".
What is the correct grant type to use?

**Comment by StuFrankish** [comment]
Little bit further - Through trial and error I've set the allowed grant type to use "authorization_code"
My provider recognises the client and allows me to enter and submit credentials, however the passback to Wiki fails with the error message "Opps, something went wrong... failed to obtain access token"

**Comment by StuFrankish** [comment]
@NGPixel - What diagnostic options are there for fault finding here?
Our identity provider currently serves a number of different applications without issue, but after authenticating and returning to the Wiki, we are presented with the error above.

There is a `500 - Internal Server Error` being thrown on;
_https://wiki.[ourwebsite].io/login/8640fd1e-4b1d-4043-8c99-f4b819552502/callback?_

There are parameters being provided to the endpoint;
**code**: c4bc24271db5601230a8f47435740f73eeeeb41538d5bda9173d7386ad90c
**scope**: openid profile email
**state**: /n60jER1hcQXwM+X7aCJ/X
**session_state**: ij1kLgPhQigXePNZsxLC18ryLZlVNlYZtdCLEAEqpmg.a0fdddd0246080174dcb1acf6d252

I have checked the container console (which proved useful in my last issue) but no errors are logged.

**Comment by Gdwalmsley** [comment]
I am having the same issue as you Saldash, my identityserver has successfull signed me in however on returning to wiki, its returning failed to obtain access token (status code 500)

Any help would be appreciated.

**Comment by olavaukan** [comment]
I might have some useful input to this issue.

I had similar problems with the error 'Failed to obtain access token' and no hint of what was going wrong. The error comes from the underlying passport-openidconnect package. You can get some hint of what is going wrong by editing the node_modules\passport-openidconnect\lib\strategy.js file to include some logging.

```
      oauth2.getOAuthAccessToken(code, { grant_type: 'authorization_code', redirect_uri: callbackURL }, function(err, accessToken, refreshToken, params) {
        if (err) {
          console.warn("Failed to obtain access token: ", err);
          return self.error(self.error('Failed to obtain access token', err));
        }
```
This didn't help me that much so I kept adding console.warning() at various places util I found the underlying issue. I found my problem in the stored() function.

```
      function stored(err, state) {
        if (err) { return self.error(err); }
        if (!state) { return self.error(new Error('Unable to generate required state parameter')); }

        params.state = state;
        var location = config.authorizationURL + '?' + querystring.stringify(params);
        console.warn("self.redirect(location): ", location);
        self.redirect(location);
      }
```
This printed the following to the console:

```
self.redirect(location):  https://dev-identity.mydomain.com/connect/authorize?response_type=code&client_id=Wiki&redirect_uri=https%3A%2F%2Fdev-docs.mydomain.com%2Flogin%2F029f03fd-9646-4b67-919b-e9f71e79008d%2Fcallback&scope=openid%20openid%20profile%20email&state=kNpjSYFokrE69h4Tf0Wyy%2FMr
Failed to obtain access token:  Error: connect ECONNREFUSED 127.0.0.1:443
    at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1146:16) {
  errno: -111,
  code: 'ECONNREFUSED',
  syscall: 'connect',
  address: '127.0.0.1',
  port: 443
}
err.stack Error: connect ECONNREFUSED 127.0.0.1:443
    at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1146:16)
```

The problem (in my case) was a combination of docker, reverse-proxy and non-public domain. I was running everything in docker with a reverse-proxy routing https://dev-identity.mydomain.com to http://wiki:3000 (Wiki.js in a container) and I had added a record to my hosts file on the host machine to be able to resolve dev-identity.mydomain.com on my machine and route that to the reverse-proxy. But Wiki.js running inside the container is not able to resolve dev-identity.mydomain.com so it's not able to redirect to the authorize-endoint and the whole thing fails. The solution is to tell the container how to resolve the domain, so for docker-compose that means adding an extra_hosts record, see https://docs.docker.com/compose/compose-file/#extra_hosts

So my particular issue was quite obscure, but I believe messing around with the strategy.js file to add logging will help you resolve your own issue.

**Reply by jensfriisnielsen** [reply]
> I might have some useful input to this issue.
> 
> I had similar problems with the error 'Failed to obtain access token' and no hint of what was going wrong. The error comes from the underlying passport-openidconnect package. You can get some hint of what is going wrong by editing the node_modules\passport-openidconnect\lib\strategy.js file to include some logging.
> 
> ```
>       oauth2.getOAuthAccessToken(code, { grant_type: 'authorization_code', redirect_uri: callbackURL }, function(err, accessToken, refreshToken, params) {
>         if (err) {
>           console.warn("Failed to obtain access token: ", err);
>           return self.error(self.error('Failed to obtain access token', err));
>         }
> ```
> 
> This didn't help me that much so I kept adding console.warning() at various places util I found the underlying issue. I found my problem in the stored() function.
> 
> ```
>       function stored(err, state) {
>         if (err) { return self.error(err); }
>         if (!state) { return self.error(new Error('Unable to generate required state parameter')); }
> 
>         params.state = state;
>         var location = config.authorizationURL + '?' + querystring.stringify(params);
>         console.warn("self.redirect(location): ", location);
>         self.redirect(location);
>       }
> ```
> 
> This printed the following to the console:
> 
> ```
> self.redirect(location):  https://dev-identity.mydomain.com/connect/authorize?response_type=code&client_id=Wiki&redirect_uri=https%3A%2F%2Fdev-docs.mydomain.com%2Flogin%2F029f03fd-9646-4b67-919b-e9f71e79008d%2Fcallback&scope=openid%20openid%20profile%20email&state=kNpjSYFokrE69h4Tf0Wyy%2FMr
> Failed to obtain access token:  Error: connect ECONNREFUSED 127.0.0.1:443
>     at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1146:16) {
>   errno: -111,
>   code: 'ECONNREFUSED',
>   syscall: 'connect',
>   address: '127.0.0.1',
>   port: 443
> }
> err.stack Error: connect ECONNREFUSED 127.0.0.1:443
>     at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1146:16)
> ```
> 
> The problem (in my case) was a combination of docker, reverse-proxy and non-public domain. I was running everything in docker with a reverse-proxy routing https://dev-identity.mydomain.com to http://wiki:3000 (Wiki.js in a container) and I had added a record to my hosts file on the host machine to be able to resolve dev-identity.mydomain.com on my machine and route that to the reverse-proxy. But Wiki.js running inside the container is not able to resolve dev-identity.mydomain.com so it's not able to redirect to the authorize-endoint and the whole thing fails. The solution is to tell the container how to resolve the domain, so for docker-compose that means adding an extra_hosts record, see https://docs.docker.com/compose/compose-file/#extra_hosts
> 
> So my particular issue was quite obscure, but I believe messing around with the strategy.js file to add logging will help you resolve your own issue.

This comment is still relevant today. You just fixed 1 days worth of headache for me trying all kinds of stuff. It doesn't help that logging currently is not configurable. I love that you just took it upon yourself to add it. Hats off to you sir
